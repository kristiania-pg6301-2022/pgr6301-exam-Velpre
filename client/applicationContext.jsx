import React from "react";
import { fetchJSON, postJSON } from "./lib/http";

export const ApplicationContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },

  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async listArticles() {
    return fetchJSON(`/api/articles`);
  },

  async createArticle(article) {
    return await postJSON("/api/articles/new", article);
  },

  async deleteArticle(article) {
    return await postJSON("/api/articles/delete", article);
  },

  async updateArticle(article) {
    return await postJSON("/api/articles/update", article);
  },

  async logout(url, reload, navigate) {
    await fetch(url, { method: "delete" });
    reload();
    navigate("/");
  },
});
