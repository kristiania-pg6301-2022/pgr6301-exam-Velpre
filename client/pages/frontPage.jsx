import { Link } from "react-router-dom";
import React from "react";

export function FrontPage() {
    //skal liste bare titell fra DB med en gang
  return (
    <div>
      <h1>VÃ¥re Artikler</h1>
      <ul>
        <li>
          <Link to={"/movies"}>List movies</Link>
        </li>
      </ul>
    </div>
  );
}
