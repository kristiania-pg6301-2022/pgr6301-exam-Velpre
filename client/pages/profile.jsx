import React from "react";
import { useNavigate } from "react-router-dom";

export function Profile({ user, reload, logout }) {

  const navigate = useNavigate();

  async function handleLogout() {
    await logout("/api/login", reload, navigate);
  }

  return (
    <div className="profile">
        <h1>Profile</h1>
      {user.google ? (
        <div>
          <h1>
            User profile: {user.google.name} ({user.google.email})
          </h1>
          <img
            src={user.google.picture}
            alt={user.google.name + " profile picture"}
          />
        </div>
      ) : (
        <h1>
          User profile: {user.hk.name} ({user.hk.email})
        </h1>
      )}
      <div>
        <button className="btn" onClick={handleLogout}>Log out</button>
      </div>
    </div>
  );
}
