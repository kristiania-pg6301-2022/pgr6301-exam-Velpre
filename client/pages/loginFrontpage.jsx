import { randomString } from "../lib/randomString";
import { sha256 } from "../lib/sha256";
import React, { useContext, useEffect, useState } from "react";
import {Route, Routes, useNavigate, useParams} from "react-router-dom";
import { ApplicationContext } from "../applicationContext";

function LoginButton({ config, label, provider }) {
    async function handleLogin() {
        const {
            authorization_endpoint,
            response_type,
            scope,
            client_id,
            code_challenge_method,
        } = config[provider];

        const state = randomString(50);
        window.sessionStorage.setItem("expected_state", state);

        const parameters = {
            response_type,
            response_mode: "fragment",
            client_id,
            state,
            scope,
            redirect_uri: `${window.location.origin}/login/${provider}/callback`,
        };

        if (code_challenge_method) {
            const code_verifier = randomString(50);
            window.sessionStorage.setItem("code_verifier", code_verifier);
            parameters.code_challenge_method = code_challenge_method;
            parameters.code_challenge = await sha256(code_verifier);
            parameters.domain_hint = "egms.no";
        }

        window.location.href =
            authorization_endpoint + "?" + new URLSearchParams(parameters);
    }

    return (
        <div>
            <button onClick={handleLogin}>{label}</button>
        </div>
    );
}

export function LoginCallback({ reload, config }) {
    const { provider } = useParams();
    const [error, setError] = useState();
    const navigate = useNavigate();
    const { registerLogin } = useContext(ApplicationContext);
    useEffect(async () => {
        const { access_token, error, error_description, state, code } =
            Object.fromEntries(
                new URLSearchParams(window.location.hash.substring(1))
            );

        const expected_state = window.sessionStorage.getItem("expected_state");
        if (!state || expected_state !== state) {
            setError("Unexpected state");
            return;
        }

        if (error || error_description) {
            setError(`Error: ${error} (${error_description})`);
            return;
        }

        if (code) {
            const { client_id, token_endpoint } = config[provider];
            const code_verifier = window.sessionStorage.getItem("code_verifier");
            const redirect_uri = `${window.location.origin}/login/${provider}/callback`;
            //Denne m√• kanskje endres
            const payload = {
                grant_type: "authorization_code",
                code,
                client_id,
                code_verifier,
                redirect_uri,
            };
            const res = await fetch(token_endpoint, {
                method: "POST",
                body: new URLSearchParams(payload),
            });
            if (!res.ok) {
                setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
                return;
            }
            const { access_token } = await res.json();
            await registerLogin(provider, { access_token });
            reload();
            navigate("/front-db");
            return;
        }

        if (!access_token) {
            setError("Missing access_token");
            return;
        }

        await registerLogin(provider, { access_token });
        reload();
        navigate("/front-db");
    }, []);

    if (error) {
        return (
            <div>
                <h1>Error</h1>
                <div>{error.toString()}</div>
            </div>
        );
    }

    return <h1>Please wait...</h1>;
}


function StartLogin({ config }) {
    return (
        <div>
            <h1>Login</h1>
            <LoginButton
                label={"Login with Google"}
                config={config}
                provider={"google"}
            />
            <LoginButton
                label={"Login with ID-porten"}
                config={config}
                provider={"hk"}
            />
        </div>
    );
}

export function LoginFrontPage({ config, reload }) {
    return (
        <Routes>
            <Route path={"/"} element={<StartLogin config={config} />} />
            <Route
                path={"/:provider/callback"}
                element={<LoginCallback config={config} reload={reload} />}
            />
            <Route path={"*"} element={<StartLogin config={config} />} />
        </Routes>
    );
}

//<Route path={"/endsession"} element={<EndSession reload={reload} />} /> - TRENGER VI DENNE
