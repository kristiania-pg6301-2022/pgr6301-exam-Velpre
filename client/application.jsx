import React, {useContext, useEffect, useState} from "react";
import { ApplicationContext } from "./applicationContext";
import { useLoading } from "./lib/useLoading";
import {
  BrowserRouter,
  Link,
  useNavigate,
  Route,
  Routes,
} from "react-router-dom";
import { LoginCallback } from "./pages/loginFrontpage";
import { LoginFrontPage } from "./pages/loginFrontpage";
import { Profile } from "./pages/profile";
import { FrontPage } from "./frontPage";

function LoginActions({ user, logout, reload }) {
  const navigate = useNavigate();
  if (!user || Object.keys(user).length === 0) {
    return (
      <button className="btn" onClick={() => navigateLogin()}>
        Login
      </button>
    );
  }

  function navigateProfile() {
    navigate("/profile");
  }

  function navigateLogin() {
    navigate("/login");
  }

  return (
    <>
      <button className="btn" onClick={() => navigateProfile()}>
        {user.google?.name ? `User profile` : `Editor profile`}
      </button>
      <button
        className="btn"
        onClick={async () => await logout("/api/login", reload, navigate)}
      >
        Log ut
      </button>
    </>
  );
}
//Array for WS
const initialMessages = [];

export function Application() {
    //Websockets
    const [articles, setArticles] = useState(initialMessages);
    const [ws, setWs] = useState();

    useEffect(() => {
        const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
        ws.onmessage = (event) => {
            const {title, plot, category} = JSON.parse(event.data);
            setArticles((articles) => [...articles, { title, category, plot }]);
        };
        setWs(ws);
    }, []);


    function handleNewArticle(articles) {
        ws.send(JSON.stringify(articles));
    }

  //Server
  const { fetchLogin, logout } = useContext(ApplicationContext);
  const { data, error, loading, reload } = useLoading(fetchLogin);

  if (error) {
    return <div>Error: {error.toString()}</div>;
  }
  if (loading) {
    return <div>Please wait...</div>;
  }

  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>Front page</Link>
        <LoginActions user={data?.user} logout={logout} reload={reload} />
      </header>

      <Routes>
        <Route
          path={"/"}
          element={<FrontPage handleNewArticle={handleNewArticle} articles={articles} user={data.user} reload={reload} />}
        />
        <Route
          path={"/login/*"}
          element={<LoginFrontPage config={data.config} reload={reload} />}
        />
        <Route
          path={"/login/:provider/callback"}
          element={<LoginCallback config={data.config} reload={reload} />}
        />

        <Route
          path={"/profile"}
          element={
            <Profile user={data?.user} reload={reload} logout={logout} />
          }
        />
      </Routes>
    </BrowserRouter>
  );
}
