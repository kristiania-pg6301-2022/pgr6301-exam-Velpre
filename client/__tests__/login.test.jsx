import { act, Simulate } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { LoginFrontPage, StartLogin } from "../pages/loginFrontpage";
import { Application } from "../application";
import { ApplicationContext } from "../applicationContext";

describe("login page", () => {
  it("redirect to log in with google", async () => {
    // replace window.location to be able to detect redirects
    const location = new URL("https://www.example.com");
    delete window.location;
    window.location = new URL(location);

    const authorization_endpoint = `https://foo.example.com/auth`;
    const client_id = `1095582733852-smqnbrhcoiasjjg8q28u0g1k3nu997b0.apps.googleusercontent.com`;

    const domElement = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <LoginFrontPage
          config={{
            google: { authorization_endpoint, client_id },
          }}
        />
      </MemoryRouter>,
      domElement
    );
    await act(async () => {
      await Simulate.click(domElement.querySelector("button"));
    });
    const redirect_uri = `${location.origin}/login/google/callback`;
    expect(window.location.origin + window.location.pathname).toEqual(
      authorization_endpoint
    );
    const params = Object.fromEntries(
      new URLSearchParams(window.location.search.substring(1))
    );
    expect(params).toMatchObject({ client_id, redirect_uri });
  });
/*
  it("posts received token to server", async () => {
    window.sessionStorage.setItem("expected_state", "test");
    // replace window.location to simulate returning
    const access_token = `abc`;
    const location = new URL(
      `https://www.example.com#access_token=${access_token}&state=test`
    );
    delete window.location;
    window.location = new URL(location);

    const domElement = document.createElement("div");
    const registerLogin = jest.fn();
    const reload = jest.fn();
    await act(() => {
      ReactDOM.render(
        <MemoryRouter initialEntries={["/google/callback"]}>
          <ApplicationContext.Provider value={{ registerLogin }}>
            <LoginFrontPage reload={reload} />
          </ApplicationContext.Provider>
        </MemoryRouter>,
        domElement
      );
    });
    expect(registerLogin).toBeCalledWith("google", { access_token });
  });

 */

  it("render Login", () => {
    const element = document.createElement("div");
    ReactDOM.render(<StartLogin />, element);
    expect(element.querySelector("h1").innerHTML).toEqual("Login");
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("can navigate to user page", async () => {
    const container = document.createElement("div");
    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Application />
        </MemoryRouter>,
        container
      );
    });
  });
});
