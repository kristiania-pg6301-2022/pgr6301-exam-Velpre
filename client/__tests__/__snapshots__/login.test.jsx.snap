// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login page render Login 1`] = `"<div class=\\"login-frontpage\\"><h1>Login</h1><div><button>User Login</button></div><div><button>Editor Login</button></div></div>"`;
